In [1]:
%tensorflow_version 1.x

TensorFlow 1.x selected.


In [2]:
import tensorflow
print(tensorflow.__version__)

1.15.2


In [3]:
# !git clone https://github.com/ildoonet/tf-pose-estimation.git > /dev/null
# %cd /content/tf-pose-estimation
# !pip -q install -r requirements.txt
# !python setup.py install

In [4]:
import argparse
import logging
import sys
import time

from tf_pose import common
import cv2
import numpy as np
from tf_pose.estimator import TfPoseEstimator
from tf_pose.networks import get_graph_path, model_wh
import matplotlib.pyplot as plt
%matplotlib inline

The TensorFlow contrib module will not be included in TensorFlow 2.0.
For more information, please see:
  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md
  * https://github.com/tensorflow/addons
  * https://github.com/tensorflow/io (for I/O related ops)
If you depend on functionality not listed there, please file an issue.




In [5]:
size = '432x368'
model = 'mobilenet_thin'

In [6]:
w, h = model_wh(size)
if w == 0 or h == 0:
    e = TfPoseEstimator(get_graph_path(model), target_size=(432, 368))
else:
    e = TfPoseEstimator(get_graph_path(model), target_size=(w, h))

[2020-11-17 19:04:04,731] [TfPoseEstimator] [INFO] loading graph from /usr/local/lib/python3.6/dist-packages/tf_pose-0.1.1-py3.6-linux-x86_64.egg/tf_pose_data/graph/mobilenet_thin/graph_opt.pb(default size=432x368)
2020-11-17 19:04:04,731 INFO loading graph from /usr/local/lib/python3.6/dist-packages/tf_pose-0.1.1-py3.6-linux-x86_64.egg/tf_pose_data/graph/mobilenet_thin/graph_opt.pb(default size=432x368)


2020-11-17 19:04:04,734 WARNING From /usr/local/lib/python3.6/dist-packages/tf_pose-0.1.1-py3.6-linux-x86_64.egg/tf_pose/estimator.py:311: The name tf.gfile.GFile is deprecated. Please use tf.io.gfile.GFile instead.



2020-11-17 19:04:04,740 WARNING From /usr/local/lib/python3.6/dist-packages/tf_pose-0.1.1-py3.6-linux-x86_64.egg/tf_pose/estimator.py:312: The name tf.GraphDef is deprecated. Please use tf.compat.v1.GraphDef instead.



2020-11-17 19:04:04,776 WARNING From /usr/local/lib/python3.6/dist-packages/tf_pose-0.1.1-py3.6-linux-x86_64.egg/tf_pose/estimator.py:330: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.



2020-11-17 19:04:04,916 WARNING From /usr/local/lib/python3.6/dist-packages/tf_pose-0.1.1-py3.6-linux-x86_64.egg/tf_pose/estimator.py:332: The name tf.Session is deprecated. Please use tf.compat.v1.Session instead.



TfPoseEstimator/MobilenetV1/Conv2d_0/weights
TfPoseEstimator/image
TfPoseEstimator/MobilenetV1/Conv2d_0/Conv2D
TfPoseEstimator/MobilenetV1/Conv2d_0/Conv2D_bn_offset
TfPoseEstimator/MobilenetV1/Conv2d_0/BatchNorm/FusedBatchNorm
TfPoseEstimator/MobilenetV1/Conv2d_0/Relu
TfPoseEstimator/MobilenetV1/Conv2d_1_depthwise/depthwise_weights
TfPoseEstimator/MobilenetV1/Conv2d_1_pointwise/weights
TfPoseEstimator/MobilenetV1/Conv2d_1_depthwise/depthwise
TfPoseEstimator/MobilenetV1/Conv2d_1_pointwise/Conv2D
TfPoseEstimator/MobilenetV1/Conv2d_1_pointwise/Conv2D_bn_offset
TfPoseEstimator/MobilenetV1/Conv2d_1_pointwise/BatchNorm/FusedBatchNorm
TfPoseEstimator/MobilenetV1/Conv2d_1_pointwise/Relu
TfPoseEstimator/MobilenetV1/Conv2d_2_depthwise/depthwise_weights
TfPoseEstimator/MobilenetV1/Conv2d_2_pointwise/weights
TfPoseEstimator/MobilenetV1/Conv2d_2_depthwise/depthwise
TfPoseEstimator/MobilenetV1/Conv2d_2_pointwise/Conv2D
TfPoseEstimator/MobilenetV1/Conv2d_2_pointwise/Conv2D_bn_offset
TfPoseEstimator/M

2020-11-17 19:04:05,120 WARNING From /usr/local/lib/python3.6/dist-packages/tf_pose-0.1.1-py3.6-linux-x86_64.egg/tf_pose/estimator.py:341: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.



2020-11-17 19:04:05,125 WARNING From /usr/local/lib/python3.6/dist-packages/tf_pose-0.1.1-py3.6-linux-x86_64.egg/tf_pose/estimator.py:342: The name tf.image.resize_area is deprecated. Please use tf.compat.v1.image.resize_area instead.



2020-11-17 19:04:05,160 WARNING From /usr/local/lib/python3.6/dist-packages/tf_pose-0.1.1-py3.6-linux-x86_64.egg/tf_pose/tensblur/smoother.py:96: The name tf.variable_scope is deprecated. Please use tf.compat.v1.variable_scope instead.



Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where


2020-11-17 19:04:05,201 WARNING From /usr/local/lib/python3.6/dist-packages/tf_pose-0.1.1-py3.6-linux-x86_64.egg/tf_pose/estimator.py:354: where (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where


2020-11-17 19:04:05,210 WARNING From /usr/local/lib/python3.6/dist-packages/tf_pose-0.1.1-py3.6-linux-x86_64.egg/tf_pose/estimator.py:359: The name tf.variables_initializer is deprecated. Please use tf.compat.v1.variables_initializer instead.



2020-11-17 19:04:05,215 WARNING From /usr/local/lib/python3.6/dist-packages/tf_pose-0.1.1-py3.6-linux-x86_64.egg/tf_pose/estimator.py:360: The name tf.global_variables is deprecated. Please use tf.compat.v1.global_variables instead.



2020-11-17 19:04:05,220 WARNING From /usr/local/lib/python3.6/dist-packages/tf_pose-0.1.1-py3.6-linux-x86_64.egg/tf_pose/estimator.py:362: The name tf.report_uninitialized_variables is deprecated. Please use tf.compat.v1.report_uninitialized_variables instead.



In [7]:
import os
import pandas as pd

In [9]:

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [10]:

os.chdir("/content/drive")

In [11]:

os.getcwd()

'/content/drive'

In [12]:
for root, dirs, files in os.walk("/content/drive", topdown=False):
   for name in files:
      print(os.path.join(root, name))
   for name in dirs:
      print(os.path.join(root, name))

/content/drive/MyDrive/USF/VISA PROCESS - ELABORATED.pdf
/content/drive/MyDrive/USF/Statement_of_Purpose.docx
/content/drive/MyDrive/USF/Statement_of_Purpose.pdf
/content/drive/MyDrive/USF/Superuser_380773_1366188.zip
/content/drive/MyDrive/USF/UCONN_ACCEPTANCE.pdf
/content/drive/MyDrive/USF/UCONN_DECISION.pdf
/content/drive/MyDrive/USF/USF_Application.pdf
/content/drive/MyDrive/USF/UTD_AD_AAA.pdf
/content/drive/MyDrive/USF/UTSCHOLARSHIP.pdf
/content/drive/MyDrive/USF/VISA_SCANNED.pdf
/content/drive/MyDrive/USF/Tuition and Fees.pdf
/content/drive/MyDrive/USF/1366188_RL.pdf
/content/drive/MyDrive/USF/Affidavit of Support.pdf
/content/drive/MyDrive/USF/1366188_SC.pdf
/content/drive/MyDrive/USF/Apply for a U.S. Visa _ Receipt for Payment.pdf
/content/drive/MyDrive/USF/AppointmentConfirmationVISA.pdf
/content/drive/MyDrive/USF/approval__letter.docx
/content/drive/MyDrive/USF/approval_letter.docx
/content/drive/MyDrive/USF/DS160-APPLICATION.pdf
/content/drive/MyDrive/USF/DS160-CONFIRMATION.

In [13]:

cutDir = '/content/drive/My Drive/training_set/Upper_Cut'
cutimages = os.listdir(cutDir)
print(cutimages)

['IMG-20201107-WA0013.jpg', 'IMG-20201107-WA0061.jpg', 'download (3).jpg', 'IMG-20201107-WA0075.jpg', 'download.jpg', 'IMG-20201107-WA0030.jpg', 'IMG-20201107-WA0040.jpg', 'download (1).jpg', 'IMG-20201107-WA0032.jpg', 'IMG-20201107-WA0060.jpg', 'download (2).jpg', 'IMG-20201107-WA0078.jpg', 'IMG-20201107-WA0022.jpg', 'download (4).jpg', 'IMG-20201107-WA0041.jpg', 'IMG-20201107-WA0044.jpg', 'images.jpg', 'IMG-20201107-WA0065.jpg', 'download (5).jpg', 'IMG-20201107-WA0034.jpg', 'IMG-20201107-WA0025.jpg', 'IMG-20201107-WA0052.jpg']


In [14]:

cutShots = []
cutFiles=[]
for img in cutimages:
    image = common.read_imgfile(cutDir+'/'+img, None, None)
    humans = e.inference(npimg=image, resize_to_default=(w > 0 and h > 0), upsample_size=4.0)
    if len(humans)>0:
        cutShots.append(humans[0])
        cutFiles.append(image)
print("Cut Shot Examples: ",len(cutFiles))

Cut Shot Examples:  20


In [15]:

sweepDir = '/content/drive/My Drive/training_set/Sweep_Shot/'
sweepimages = os.listdir(sweepDir)

In [16]:

sweepShots = []
sweepFiles=[]
for img in sweepimages:
    image = common.read_imgfile(sweepDir+img, None, None)
    humans = e.inference(npimg=image, resize_to_default=(w > 0 and h > 0), upsample_size=4.0)
    if len(humans)>0:
        sweepShots.append(humans[0])
        sweepFiles.append(image)
print("Sweep  Examples: ",len(sweepFiles))

Sweep  Examples:  42


In [17]:
coverDir = '/content/drive/My Drive/training_set/Cover_Drive'
coverimages = os.listdir(coverDir)

In [18]:
coverShots = []
coverFiles=[]
for img in coverimages:
    image = common.read_imgfile(coverDir+'/'+img, None, None)
    humans = e.inference(npimg=image, resize_to_default=(w > 0 and h > 0), upsample_size=4.0)
    if len(humans)>0:
        coverShots.append(humans[0])
        coverFiles.append(image)
print("cover Examples:",len(coverFiles))

cover Examples: 39


In [19]:
pullDir = '/content/drive/My Drive/training_set/Pull_Shot'
pullimages = os.listdir(pullDir)

In [20]:
pullShots = []
pullFiles=[]
for img in pullimages:
    image = common.read_imgfile(pullDir+'/'+img, None, None)
    humans = e.inference(npimg=image, resize_to_default=(w > 0 and h > 0), upsample_size=4.0)
    if len(humans)>0:
        pullShots.append(humans[0])
        pullFiles.append(image)
print("Pull Examples: ",len(pullFiles))

Pull Examples:  39


In [21]:
def humanToDict(hum):
    resultDict = {}
    parts = hum.body_parts.keys()
    for p in parts:
        resultDict[str(p)+'_x'] = hum.body_parts[p].x
        resultDict[str(p)+'_y'] = hum.body_parts[p].y
        #resultDict[str(p)+'_p'] = hum.body_parts[p].p
    return resultDict

In [22]:
cutList = []
for sh in cutShots:
    cutList.append(humanToDict(sh))
    
cutdf = pd.DataFrame(cutList)
#standHumadf['img'] = standFiles
cutdf.head()

1_x       1_y       2_x       2_y  ...      17_x      17_y  7_x  7_y
0  0.347222  0.396739  0.314815  0.423913  ...       NaN       NaN  NaN  NaN
1  0.564815  0.342391  0.643519  0.342391  ...       NaN       NaN  NaN  NaN
2  0.370370  0.369565  0.324074  0.375000  ...  0.379630  0.326087  NaN  NaN
3  0.449074  0.260870  0.537037  0.239130  ...  0.388889  0.222826  NaN  NaN
4  0.750000  0.336957       NaN       NaN  ...  0.782407  0.277174  NaN  NaN

[5 rows x 36 columns]

In [23]:
sweepList = []
for jh in sweepShots:
    sweepList.append(humanToDict(jh))

sweepdf = pd.DataFrame(sweepList)
sweepdf.head()

0_x       0_y       1_x  ...      17_y      16_x      16_y
0  0.314815  0.168478  0.379630  ...  0.130435       NaN       NaN
1  0.194444  0.336957  0.231481  ...  0.288043  0.148148  0.326087
2  0.402778  0.157609  0.412037  ...  0.135870  0.347222  0.141304
3  0.810185  0.228261  0.847222  ...  0.250000       NaN       NaN
4  0.254630  0.271739  0.287037  ...  0.222826  0.208333  0.293478

[5 rows x 36 columns]

In [24]:
driveList = []
for jh in coverShots:
    driveList.append(humanToDict(jh))

drivedf = pd.DataFrame(driveList)
drivedf.head()

1_x       1_y       2_x  ...      10_y      16_x      16_y
0  0.305556  0.157609  0.314815  ...       NaN       NaN       NaN
1  0.689815  0.179348  0.592593  ...       NaN       NaN       NaN
2  0.490741  0.222826  0.393519  ...       NaN       NaN       NaN
3  0.638889  0.250000  0.550926  ...       NaN       NaN       NaN
4  0.439815  0.217391  0.486111  ...  0.880435  0.476852  0.168478

[5 rows x 36 columns]

In [25]:
pullList = []
for jh in pullShots:
    pullList.append(humanToDict(jh))

pulldf = pd.DataFrame(pullList)
pulldf.head()

0_x       0_y       1_x  ...      13_y      17_x      17_y
0  0.472222  0.168478  0.379630  ...       NaN       NaN       NaN
1  0.490741  0.125000  0.472222  ...       NaN       NaN       NaN
2  0.398148  0.190217  0.398148  ...  0.809783  0.425926  0.146739
3  0.782407  0.190217  0.787037  ...  0.445652  0.791667  0.184783
4  0.324074  0.152174  0.282407  ...       NaN  0.356481  0.135870

[5 rows x 36 columns]

In [26]:
sweepdf['pose']=0
cutdf['pose']=1
drivedf['pose']=2
pulldf['pose']=3

alldata = sweepdf.append(cutdf)
alldata = alldata.append(drivedf)
alldata = alldata.append(pulldf)

alldata = alldata.reset_index(drop = True)
alldata.shape

(140, 37)

In [27]:
allFiles = sweepFiles + cutFiles + coverFiles + pullFiles
allShots = sweepShots + cutShots + coverShots + pullShots

In [28]:
def min_max_normalize(df):
    xcols = [c for c in df.columns if 'x' in c]
    xdf = df[xcols]
    xdf = xdf.subtract(xdf.min(axis=1), axis=0)
    xdf = xdf.divide(xdf.max(axis=1), axis=0)
    
    ycols = [c for c in df.columns if 'y' in c]
    ydf = df[ycols]
    ydf = ydf.subtract(ydf.min(axis=1), axis=0)
    ydf = ydf.divide(ydf.max(axis=1), axis=0)
    if 'pose' in df.columns:
        resultdf =  pd.concat( [xdf,ydf, df[['pose']]], axis=1 )
    else:
        resultdf =  pd.concat( [xdf,ydf], axis=1 )
    return resultdf

In [29]:
alldata = min_max_normalize(alldata)

**Model 1** -> MLP Classifier



In [71]:
X = treatData[[c for c in treatData.columns if c != 'pose']]
X['img'] = allFiles
X['human'] = allShots
Y = treatData['pose']

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.3, random_state=2019, shuffle=True)

testImg = X_test.img
testHuman = X_test.human
X_train = np.array(X_train[[c for c in X_train.columns if c not in ['img','human'] ]]) 
X_test = np.array(X_test[[c for c in X_test.columns if c not in ['img','human'] ]]) 

print(X_train.shape)

from sklearn.neural_network import MLPClassifier

mlp_clf = MLPClassifier(hidden_layer_sizes=(150,100,50), max_iter=300,activation = 'relu',solver='adam',random_state=1)

mlp_clf.fit(X_train, y_train)

(98, 36)


MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(150, 100, 50), learning_rate='constant',
              learning_rate_init=0.001, max_fun=15000, max_iter=300,
              momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
              power_t=0.5, random_state=1, shuffle=True, solver='adam',
              tol=0.0001, validation_fraction=0.1, verbose=False,
              warm_start=False)

In [72]:
test_y_pred = mlp_clf.predict(X_test)

In [73]:
from sklearn.metrics import accuracy_score

print(accuracy_score(y_test, test_y_pred))

0.5


This model has accuracy of 50%

**Model 2** -> CNN

In [74]:
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Dense, Flatten, Dropout
from keras.optimizers import Adam
from keras.callbacks import TensorBoard

In [95]:
cnn_model = Sequential()

# Try 32 fliters first then 64
cnn_model.add(Conv2D(32, (3, 1), input_shape = (36,1,1), activation='relu'))

cnn_model.add(Dropout(0.25))

cnn_model.add(Conv2D(64, (3, 1), input_shape = (36,1,1), activation='relu'))

cnn_model.add(Dropout(0.25))

cnn_model.add(Conv2D(128, (3, 1), input_shape = (36,1,1), activation='relu'))

cnn_model.add(Dropout(0.25))


cnn_model.add(Flatten())
cnn_model.add(Dense(units = 512, activation = 'relu'))
cnn_model.add(Dropout(0.25))
cnn_model.add(Dense(units = 10, activation = 'softmax'))

In [96]:
cnn_model.compile(loss ='sparse_categorical_crossentropy', optimizer='adam' ,metrics =['accuracy'])

In [97]:
X_train = X_train.reshape(X_train.shape[0], *(36, 1, 1))
X_test = X_test.reshape(X_test.shape[0], *(36, 1, 1))

In [98]:
epochs = 15

history = cnn_model.fit(X_train, y_train, batch_size = 512, epochs = epochs,
                        verbose = 1)

Epoch 1/15
98/98 [==============================] - 0s 3ms/step - loss: 2.3427 - accuracy: 0.0510
Epoch 2/15
98/98 [==============================] - 0s 714us/step - loss: 1.9752 - accuracy: 0.3469
Epoch 3/15
98/98 [==============================] - 0s 717us/step - loss: 1.6395 - accuracy: 0.3980
Epoch 4/15
98/98 [==============================] - 0s 769us/step - loss: 1.4175 - accuracy: 0.3776
Epoch 5/15
98/98 [==============================] - 0s 812us/step - loss: 1.3259 - accuracy: 0.4490
Epoch 6/15
98/98 [==============================] - 0s 820us/step - loss: 1.2173 - accuracy: 0.5102
Epoch 7/15
98/98 [==============================] - 0s 758us/step - loss: 1.2016 - accuracy: 0.5102
Epoch 8/15
98/98 [==============================] - 0s 837us/step - loss: 1.2295 - accuracy: 0.4592
Epoch 9/15
98/98 [==============================] - 0s 823us/step - loss: 1.1827 - accuracy: 0.4898
Epoch 10/15
98/98 [==============================] - 0s 806us/step - loss: 1.0722 - accuracy: 0.5612
E

In [99]:
evaluation = cnn_model.evaluate(X_test, y_test)
print(f'Test Accuracy : {evaluation[1]:.3f}')

42/42 [==============================] - 0s 2ms/step
Test Accuracy : 0.524


This model has accuracy of 52.4% which is slightly better than MLP model.